addons/web_editor/models/ir_ui_view.py

```diff
@@ -270,3 +271,89 @@ def get_related_views(self, key, bundles=False):
+
+    @api.model
+    def save_snippet(self, name, arch, template_key, snippet_class=None, thumbnail_url=None):
+        """
+        Save a new snippet arch so that it appears with the given name when
+        using the given snippets template.
+
+        Params:
+            name (str): the name of the snippet to save
+
+            arch (str): the html structure of the snippet to save
+
+            template_key (str):
+                the key of the view regrouping all snippets in which the
+                snippet to save is meant to appear
+
+            snippet_class (str, default=None):
+                a className which is supposed to uniquely-identify the snippet
+                from which the snippet to save originates
+
+            thumbnail_url (str, default=None):
+                the url of the thumbnail to use when displaying the snippet to
+                save (default one: see '_get_default_snippet_thumbnail')
+        """
+        if not thumbnail_url:
+            thumbnail_url = self._get_default_snippet_thumbnail(snippet_class)
+
+        app_name = template_key.split('.')[0]
+        snippet_class = snippet_class or 's_custom_snippet'
+        key = '%s_%s' % (snippet_class, uuid.uuid4().hex)
+        snippet_key = '%s.%s' % (app_name, key)
+
+        # html to xml to add '/' at the end of self closing tags like br, ...
+        xml_arch = etree.tostring(html.fromstring(arch))
+        new_snippet_view_values = {
+            'name': name,
+            'key': snippet_key,
+            'type': 'qweb',
+            'arch': xml_arch,
+        }
+        new_snippet_view_values.update(self._snippet_save_view_values_hook(app_name))
+        self.create(new_snippet_view_values)
+
+        custom_section = self.search([('key', '=', template_key)])
+        snippet_addition_view_values = {
+            'name': name + ' Block',
+            'key': self._get_snippet_addition_view_key(template_key, key),
+            'inherit_id': custom_section.id,
+            'type': 'qweb',
+            'arch': """
+                <data inherit_id="%s">
+                    <xpath expr="//div[@id='snippet_custom']" position="attributes">
+                        <attribute name="class" remove="d-none" separator=" "/>
+                    </xpath>
+                    <xpath expr="//div[@id='snippet_custom_body']" position="inside">
+                        <t t-snippet="%s" t-thumbnail="%s"/>
+                    </xpath>
+                </data>
+            """ % (template_key, snippet_key, thumbnail_url),
+        }
+        snippet_addition_view_values.update(self._snippet_save_view_values_hook(app_name))
+        self.create(snippet_addition_view_values)
+
+    @api.model
+    def delete_snippet(self, view_id, template_key):
+        snippet = self.browse(view_id)
+        key = snippet.key.split('.')[1]
+        custom_key = self._get_snippet_addition_view_key(template_key, key)
+        self.search([('key', '=', custom_key)]).unlink()
```

Comment:
can unlink both in one sql request

Reply:
Would you do it though? ORM instead of SQL seems ok for such a small use case

Reply:
yes I always concat records then unlink in those case, when unlink are done that close, no point to fire 2 requests
```python
custom_key_view = self.search([('key', '=', custom_key)]).unlink()
(snippet | custom_key_view).unlink()
```
nothing critical tho could keep it that way

Reply:
Ok, did not think about doing it this way, it is indeed better. I'll change.